nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/repository/claim/ContentClaimWriteCache.java

```diff
@@ -0,0 +1,168 @@
+    private final Map<ResourceClaim, ByteCountingOutputStream> streamMap = new HashMap<>();
+    private final Queue<ContentClaim> queue = new LinkedList<>();
+    private final int bufferSize;
+
+    public ContentClaimWriteCache(final ContentRepository contentRepo) {
+        this(contentRepo, 8192);
+    }
+
+    public ContentClaimWriteCache(final ContentRepository contentRepo, final int bufferSize) {
+        this.contentRepo = contentRepo;
+        this.bufferSize = bufferSize;
+    }
+
+    public void reset() throws IOException {
+        try {
+            forEachStream(OutputStream::close);
+        } finally {
+            streamMap.clear();
+            queue.clear();
+        }
+    }
+
+    public ContentClaim getContentClaim() throws IOException {
+        final ContentClaim contentClaim = queue.poll();
+        if (contentClaim != null) {
+            contentRepo.incrementClaimaintCount(contentClaim);
+            return contentClaim;
+        }
+
+        final ContentClaim claim = contentRepo.create(false);
+        registerStream(claim);
+        return claim;
+    }
+
+    private ByteCountingOutputStream registerStream(final ContentClaim contentClaim) throws IOException {
+        final OutputStream out = contentRepo.write(contentClaim);
+        final OutputStream buffered = new BufferedOutputStream(out, bufferSize);
+        final ByteCountingOutputStream bcos = new ByteCountingOutputStream(buffered);
+        streamMap.put(contentClaim.getResourceClaim(), bcos);
+        return bcos;
+    }
+
+    public OutputStream write(final ContentClaim claim) throws IOException {
+        OutputStream out = streamMap.get(claim.getResourceClaim());
+        if (out == null) {
+            out = registerStream(claim);
+        }
+
+        if (!(claim instanceof StandardContentClaim)) {
+            // we know that we will only create Content Claims that are of type StandardContentClaim, so if we get anything
+            // else, just throw an Exception because it is not valid for this Repository
+            throw new IllegalArgumentException("Cannot write to " + claim + " because that Content Claim does belong to this Claim Cache");
+        }
+
+        final StandardContentClaim scc = (StandardContentClaim) claim;
+        final long initialLength = Math.max(0L, scc.getLength());
+
+        final OutputStream bcos = out;
+        return new OutputStream() {
+            private long bytesWritten = 0L;
+
+            @Override
+            public void write(final int b) throws IOException {
+                bcos.write(b);
+                bytesWritten++;
+                scc.setLength(initialLength + bytesWritten);
+            }
+
+            @Override
+            public void write(byte[] b, int off, int len) throws IOException {
+                bcos.write(b, off, len);
+                bytesWritten += len;
+                scc.setLength(initialLength + bytesWritten);
+            }
+
```

Comment:
It is correct once I make this change
```
private long bytesWritten = Math.max(0L, scc.getLength());
```
And there is an existing test to prove it. Curious though why would scc.length ever be < 0?